### PR TITLE
desktop deep linking

### LINKED ISSUE #5038: Add custom URL scheme to status-desktop
### User Story

As a user in want to open universal links directly in my status-desktop app.

### Description

*Type*: Feature

*Summary*: The goal is to eliminate any friction to opening up Status. No more copy/pasting long contact codes, or cumbersome instructions (Open Status, tap Home, then New, then ……..). 
Description of universal links is [here](https://wiki.status.im/Universal_Links). 
Pull request on mobile side: #4944 

### Solution

*Summary*: 

Clicking these links should open Status and deep link to a chat or contact

- [status-im://chat/public/status](status-im://chat/public/status)
- [status-im://user/0x044166fa04e967e333211ae0cc1087c9c74exxx](status-im://user/0x044166fa04e967e333211ae0cc1087c9c74exxx)

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #5038 

### Summary:

Adds universal links support: 

* Chat links like "status-im://chat/public/status".
* Contact links like "status-im://chat/public/0x04..." 
* DApp Browser links are not supported on desktop.

status: ready

### CODE REVIEW

File: src/status_im/events.cljs [hunk 1]
```diff
@@ -1044,3 +1046,9 @@
  [(re-frame/inject-cofx :random-id-generator)]
  (fn [cofx _]
    (contact/add-new-identity-to-contacts cofx)))
+
+(handlers/register-handler-fx
+ :desktop/handle-universal-link
+ (fn [cofx [_ url]]
+   (desktop.deep-links/on-url-opened)
```
Comment (yenda) [inline src/status_im/events.cljs]: why is this needed here ?

Comment (dmitryn) [reply src/status_im/events.cljs]: @yenda `on-url-opened` is needed to set up new callback for urlOpened event. We need to set up it each time after last callback was fired and destroyed.

File: modules/react-native-desktop-linking/desktop/desktoplinking.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+signals:
+    void urlOpened(QString path);
+    void fileOpened(QString path);
+
+public slots:
+    void handleURL(const QString url);
+
+private:
+    QScopedPointer<DesktopLinkingPrivate> d_ptr;
+    bool eventFilter(QObject* obj, QEvent* event)
```
Comment (maxhora) [inline modules/react-native-desktop-linking/desktop/desktoplinking.h]: Could you please move implementation part of `eventFilter` to `desktoplinking.cpp`

File: src/status_im/events.cljs (hunk 1 above)
Comment (maxhora) [reply src/status_im/events.cljs]: @dmitryn  Does it make more sense to listen to the custom native event all the time?
Implementation sample: NetInfo.js subscribes and listens to native custom events there https://github.com/status-im/react-native-desktop/blob/master/Libraries/Network/NetInfo.js#L19
The event is generated from the cpp code: https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/netinfo.cpp#L80

Comment (dmitryn) [reply src/status_im/events.cljs]: @MaxRis yes, it makes sense. I will use that approach. Thanks.

File: src/status_im/ui/components/desktop/events.cljs [hunk 3]
```diff
@@ -1,19 +1,37 @@
...
+            [status-im.utils.fx :as fx]
             [status-im.utils.platform :as platform]))
 
+(fx/defn change-tab
+  [{:keys [db]} tab-name]
+  {:db (assoc-in db [:desktop/desktop :tab-view-id] tab-name)})
+
+(fx/defn navigate-to
+  [{:keys [db]} tab-name]
+  {:dispatch [:navigate-to (if (and (= tab-name :home) (:current-chat-id db))
```
Comment (yenda) [inline src/status_im/ui/components/desktop/events.cljs]: Why not use navigate-to-cofx FN directly?

Comment (dmitryn) [reply src/status_im/ui/components/desktop/events.cljs]: @yenda yes, changing it to `navigate-to-cofx`

File: modules/react-native-desktop-linking/desktop/desktoplinking.h (hunk 2 above)
Comment (dmitryn) [reply modules/react-native-desktop-linking/desktop/desktoplinking.h]: @MaxRis moved and changed implementation to using `NativeEventEmitter`. Could you please review again.

**Review by flexsurfer — APPROVED** [review]

**Comment by vitvly** [comment]
Great work @dmitryn !

I've been trying it out locally. Clicking on a `status` universal link from a Status chat in a dev instance correctly switches me to the Status PR build window. But I see the error `re-matches must match against a string`, when clicking e.g. on `status-im://chat/public/status` link.

**Comment by vitvly** [comment]
This is the stacktrace from log:
```
TypeError: re-matches must match against a string.
    at Object.cljs.core.re_matches (evalmachine.<anonymous>:12:719668)
    at Object.status_im.utils.universal_links.core.match_url (evalmachine.<anonymous>:12:5361757)
    at Function.status_im.utils.universal_links.core.route_url.cljs$core$IFn$_invoke$arity$2 (evalmachine.<anonymous>:12:5373095)
    at Function.status_im.utils.universal_links.core.handle_url.cljs$core$IFn$_invoke$arity$2 (evalmachine.<anonymous>:12:5377767)
    at evalmachine.<anonymous>:12:7761786
    at evalmachine.<anonymous>:12:1549867
    at Object.re_frame.interceptor.invoke_interceptor_fn (evalmachine.<anonymous>:12:1480005)
    at Object.re_frame.interceptor.invoke_interceptors (evalmachine.<anonymous>:12:1480289)
    at Object.re_frame.interceptor.execute (evalmachine.<anonymous>:12:1482740)
    at Object.re_frame.events.handle (evalmachine.<anonymous>:12:1491165)
```

**Comment by chadyj** [comment]
Super excited for this to be in! Will making joining channels and adding contacts so much easier :)

I downloaded the Mac build but also got the `re-matches must match against a string` error.

screen shot 2018-10-17 at 1 25 54 pm

**Comment by dmitryn** [comment]
@siphiuel @chadyj it looks like you guys using old build. Can you try the last one? It's still building, but MacOS link is already there: https://ci.status.im/job/status-react/job/combined/job/desktop-macos/2632/artifact/StatusImPackage/StatusIm-181017-101805-c00366-pr.dmg

<blockquote></blockquote>

**Comment by chadyj** [comment]
Worked beautifully!

I was opening URL's from https://notes.status.im/status-channels and https://people-ops.status.im/status-direct/

I also tested logging out, clicking a link (and getting login screen), then logging in and having the channel I previously clicked show up.

Nice one 👍 

Desktop just got so much more usable :)

**Comment by vitvly** [comment]
Yes, latest build worked for me as well. Thank you @dmitryn

**Review by vitvly — APPROVED** [review]

**Review by maxhora — APPROVED** [review]
@dmitryn looks great!! Thanks!

**Comment by asemiankevich** [comment]
@dmitryn is it okay that mobile build from this PR does not contain universal links in profile (profile - share link)? develop and nightly looks good both.

**Comment by dmitryn** [comment]
@asemiankevich i haven't changed anything related to that, let me pull recent changed from develop here and check

**Comment by asemiankevich** [comment]
@dmitryn yeah super weird , maybe rebase can help

**Comment by dmitryn** [comment]
@asemiankevich build is ready, please check

**Comment by asemiankevich** [comment]
1. can open chat unversal link on desktop
2. can open profile universal link on desktop
3. dapps universal links are not working (expected here in this PR).

looks good. No mobile regressions caused.

**Comment by chadyj** [comment]
Thanks y'all! This is a great feature.

Does it work on linux? (if not, lets create a separate issue)

**Comment by asemiankevich** [comment]
i dont have linux so i did not check that @chadyj 
@dmitryn do you know?

**Comment by dmitryn** [comment]
Let me check on Linux (Ubuntu)

**Comment by dmitryn** [comment]
@chadyj doesn't work on Linux, let's create a separate issue for that.

screenshot 2018-10-18 16 00 14
